### PR TITLE
Firebase: Add maxRecordSize and maxPropertySize limits. Refactor firebase config logic.

### PR DESCRIPTION
**Pull request by davidsbailey**
This PR enforces a maximum of 4096 bytes for the JSON-encoded values of records and key-value pair values via firebase security rules. Applab reads the same limits from firebase config and catches violations before they happen so that it can fail gracefully. the 4096-byte limit was previously in place for key-value pairs and table records in our DynamoDB implementation.

minor caveat: a record of JSON-encoded size 4097 bytes which includes a record id > `1,000,000,000` would result in an opaque `Permission denied` error message and the row count for that table would become invalid, because the size of the ID would exceed the padding that applab allows for it, and the security rule would fail after `rowCount` had already been updated. This seems sufficiently improbable to me to ignore this case but I'd be happy to add any additional protections to mitigate a bad outcome.

It's also worth noting that firebase imposes a 768-character limit on the length of the key of any node, so we don't explicitly need to add our own limits on the property key length unless we want a different limit: https://firebase.google.com/docs/database/web/structure-data

```
If you create your own keys, they must be UTF-8 encoded, can be a maximum of 768 bytes, 
and cannot contain ., $, #, [, ], /, or ASCII control characters 0-31 or 127.
```

The work to track the exclusion of the forbidden characters listed above (or perhaps to UTF-8) is being tracked in a list of loose ends which I have yet to put in Pivotal. Example applab commands which would fail ungracefully are `setKeyValue('a.b', 123)` and `setKeyValue('☃', 'hello')`. As I write this I realize that we would want add logic to fail gracefully when the 768-byte limit is exceeded as well.

### CODE REVIEW

File: apps/src/applab/firebaseCounters.js [hunk 1]
```diff
@@ -76,16 +61,21 @@ export function updateTableCounters(tableName, rowCountChange, updateNextId) {
...
-  return loadConfig().then(() => getCurrentTime())
-  .then(currentTimeMs => {
-    let promises = [];
-    Object.keys(rateLimitMap).forEach(interval => {
+  let promises = [];
+  promises.push(loadConfig());
+  promises.push(getCurrentTime());
+  return Promise.all(promises).then(results => {
+    const config = results[0];
+    const currentTimeMs = results[1];
```
Comment (islemaster) [inline apps/src/applab/firebaseCounters.js]: Could use array destructuring [(mdn)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment) here:

``` javascript
const [config, currentTimeMs] = results;
```

File: apps/src/applab/firebaseCounters.js [hunk 2]
```diff
@@ -110,42 +100,10 @@ export function incrementRateLimitCounters() {
...
-  const configRef = getConfig();
-  return configRef.once('value', snapshot => {
-    handleLoadConfig(snapshot.val());
-
-    // Make sure we don't listen multiple times.
-    configRef.off();
-
-    // Update globals to reflect firebase config any time it changes in the future.
-    configRef.on('value', snapshot => handleLoadConfig(snapshot.val()));
+    return Promise.all(intervalPromises);
```
Comment (islemaster) [inline apps/src/applab/firebaseCounters.js]: You've nearly reduced this to a `map` operation.  Could:

``` javascript
return Promise.all(Object.keys(config.limits).map(interval => {
  const limitRef = getDatabase(Applab.channelId).child(`counters/limits/${interval}`);
  return limitRef.transaction(/* ... */);
}));
```

Up to you whether that would read better.

**Comment by islemaster** [comment]
LGTM.

File: apps/src/applab/firebaseCounters.js (hunk 2 above)
Comment (davidsbailey) [inline apps/src/applab/firebaseCounters.js]: I love it! I've also inlined the other `promise` array.

File: apps/src/applab/firebaseCounters.js (hunk 1 above)
Comment (davidsbailey) [inline apps/src/applab/firebaseCounters.js]: Ah, yes! I had tried that, but not found the right syntax for doing it with const.

**Comment by davidsbailey** [comment]
Thanks much for the es6 tips @islemaster ! arrow functions and promises go so well together I almost want to use them for my synchronous code ;)

**Comment by davidsbailey** [comment]
Rebasing onto staging-next and merging there via https://github.com/code-dot-org/code-dot-org/pull/9028.

**Pull request closed without merging**